### PR TITLE
feat: start root chain sync from the last app state

### LINKED ISSUE #384: Do not re-read all the bridge events on startup
# Bounty

We are [storing](https://github.com/leapdao/leap-node/pull/241/files#diff-98cc4f4c596f029c967bc7f214114543R181) tendermint app hash now, so we don't do full resync on startup [anymore](https://github.com/leapdao/leap-node/pull/241/files#diff-615e354a35574ade81a8b6d902203052R210). However, we are still re-reading all the root chain events on startup, which is totally unnecessary now ~and introduces extra complexity like keeping `epochLengths` array in a state and so on~.

## Scope

- when storing the app state, store the last root chain block read, but keep it out of consensus state (that's important)
- when starting the node, read bridge events not from genesis, but from the last root chain block persisted
- remove all the code used to support resync from genesis ~(e.g. collecting epochLengths in the array etc)~

## Deliverables

- updated code
- updated tests

## Gain for the project

- faster node starts
- simpler code, easier to reason about

## Roles
bounty gardener: @troggy / 40 DAI
bounty worker: @troggy / 75%
bounty reviewer: name / 25%

## Funded by

Dev Circle

### PR DESCRIPTION
**Pull request by troggy**
We keep the app state, so we don't need to replay everything from the genesis. However,
we kept rereading all the root chain events on each restart. Proposed change: store the last
root chain height processed and resume the sync from that height on restart.

Accompanying change:
- storing period proposals not every block, but once created (every 32 blocks)

Resolves #384

### CODE REVIEW

**Comment by troggy** [comment]
What to check:
1. On the first start it will reread everything like before (cause `lastSeenRootChainBlock` is not set yet)
2. After the next period creation, restarts shouldn't reread everything from genesis, but only from the root chain height at the time of period creation. Essentially, it replays the current period only.

File: src/validator/periods/startNewPeriod.js [hunk 1]
```diff
@@ -35,6 +35,8 @@ module.exports = async (height, bridgeState) => {
     prevPeriodRoot: bridgeState.lastPeriodRoot || GENESIS,
   };
 
+  bridgeState.savePeriodProposals();
```
Comment (sunify) [inline src/validator/periods/startNewPeriod.js]: No need to await here?

File: src/bridgeState.js [hunk 2]
```diff
@@ -299,17 +304,20 @@ module.exports = class BridgeState {
...
-    });
-
     if (!this.submissions.length) return;
     await this.db.storePeriods(this.submissions).then(() => {
       this.submissions = [];
     });
   }
 
+  async savePeriodProposals() {
+    await this.db.storeNodeState({
```
Comment (sunify) [inline src/bridgeState.js]: Unnecessary await, you can just return result of `storeNodeState`

File: src/bridgeState.js [hunk 3]
```diff
@@ -299,17 +304,20 @@ module.exports = class BridgeState {
   }
 
   async saveNodeState() {
-    await this.db.storeNodeState({
-      periodProposal: this.periodProposal,
-      stalePeriodProposal: this.stalePeriodProposal,
-    });
-
     if (!this.submissions.length) return;
     await this.db.storePeriods(this.submissions).then(() => {
```
Comment (sunify) [inline src/bridgeState.js]: Why then here? Just set `this.submissions` after `await this.db.storePeriods(this.submissions)`:

```es6
await this.db.storePeriods(this.submissions);
this.submissions = [];
```

Comment (troggy) [reply src/bridgeState.js]: probably I had brain corruption 😅

**Pull request merged**